recirq/otoc/utils.py

```diff
@@ -0,0 +1,253 @@
+            to 0 or a non-zero offset.
+
+    Returns:
+        pauli_error: The average Pauli error per cycle from the fit.
+        x_vals: A list of 200 evenly spaced values between 0 and
+            max(num_cycle_range).
+        y_vals: The fitted fidelity for each value in x_vals.
+    """
+    f_s = 1.0e2
+
+    def _exp_decay_with_offset(length: np.ndarray, err: float,
+                               a_coeff: float, b_coeff: float) -> np.ndarray:
+        p = 1.0 - err / (1.0 - 4 ** (-num_qubits)) / f_s
+        return a_coeff * np.power(p, length) + b_coeff
+
+    def _exp_decay_no_offset(length: np.ndarray, err: float,
+                             a_coeff: float) -> np.ndarray:
+        return _exp_decay_with_offset(length, err, a_coeff, 0.0)
+
+    fit_fun = _exp_decay_with_offset if add_offset else _exp_decay_no_offset
+    params_fitted, _ = curve_fit(fit_fun, num_cycle_range, data, ftol=1e-2)
+    pauli_error = float(params_fitted[0] / f_s)
+    x_vals = np.linspace(0.0, max(num_cycle_range), 200)
+    y_vals = fit_fun(x_vals, *params_fitted)
+    return pauli_error, x_vals, y_vals
+
+
+def bits_to_probabilities(all_qubits: Sequence[Tuple[int, int]],
+                          subsys_qubits: Sequence[Tuple[int, int]],
+                          bits: np.ndarray
+                          ) -> np.ndarray:
+    """Converts a collection of bit-strings into a probability distribution.
+
+    Args:
+        all_qubits: A list of Tuples representing the (row, column) locations
+            of all qubits that are measured. The qubits are assumed to be on
+            a 2D grid.
+        subsys_qubits: The qubits whose joint probabilities are to be
+            computed. They should be a subset of all_qubits.
+        bits: a 2D np.array where each row represents a bit-string (with the
+            same length as all_qubits) and each column represents a trial.
+
+    Returns:
+        The probabilities of all possible bit-strings for subsys_qubits,
+        ordered according to the big-endian notation.
+    """
+    num_qubits = len(subsys_qubits)
+    indices = [all_qubits.index(q) for q in subsys_qubits]
+    bits_red = np.asarray(bits[:, indices], dtype=int)
+    for i in range(num_qubits):
+        bits_red[:, i] *= 2 ** (num_qubits - i - 1)
+    bits = list(np.sum(bits_red, axis=1))
+    p_reduced = [bits.count(i) for i in range(2 ** num_qubits)]
+    num_trials, _ = bits_red.shape
+    return np.asarray(p_reduced) / num_trials
+
+
+def angles_to_fsim(theta: float, phi: float, delta_plus: float,
+                   delta_minus_diag: float, delta_minus_off_diag: float
+                   ) -> np.ndarray:
+    """Converts five phases to a 2x2 FSIM unitary.
+
+    Args:
+        theta: The swap angle.
+        phi: The conditional phase.
+        delta_plus: A single-qubit phase.
+        delta_minus_diag: A single-qubit phase.
+        delta_minus_off_diag: A single-qubit phase.
```

Comment:
Thanks. I will add this in the future.

Reply:
Yeah, we should use `PhasedFSimGate` as the implementation for this function or instead of this function. If you're not going to do it in this PR, please open an issue

Reply:
I opened an issue on updating Cirq to allow usage of this gate.